keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -0,0 +1,730 @@
+
+#### Bucket
+
+A cluster-scoped custom resource representing the abstraction of a single backend bucket. A `Bucket` instance stores enough identifying information so that drivers can accurately target the backend object store (e.g. needed during a deletion process).  The relevant bucket class fields are copied to the `Bucket`. Additionally, data returned by the driver is copied to the `Bucket` by the sidecar.
+
+For greenfield, COSI creates the `Bucket` based on values in the `BucketRequest` and `BucketClass`. For brownfield, an admin manually creates the `Bucket`, filling in BucketClass fields, such as `allowedNamespaces`. COSI populates fields returned by the provisioner, and binds the `Bucket` to the `BucketAccess`.
+
+Since there is a 1:1 mapping between a BR and a `Bucket`, when multiple BRs request access to the same backend bucket, each separate `Bucket` points to the same backend bucket.
+
+```yaml
+apiVersion: cosi.io/v1alpha1
+kind: Bucket
+metadata:
+  name: [1]
+  labels: [2]
+    cosi.io/provisioner:
+    cosi.io/bucket-prefix:
+  finalizers:
+    - cosi.io/finalizer [3]
+spec:
+  provisioner: [4]
+  retentionPolicy: [5]
+  anonymousAccessMode: [6]
+  bucketClassName: [7]
+  bucketRequest: [8]
+  allowedNamespaces: [9]
+    - name:
+  protocol: [10]
+    name:
+    version:
+    azureBlob:
+      containerName:
+      storageAccount:
+    s3:
+      endpoint:
+      bucketName:
+      region:
+      signatureVersion:
+    gs:
+      bucketName:
+      privateKeyName:
+      projectId:
+      serviceAccount:
+  parameters: [11]
+status:
+  bucketAvailable: [12]
+```
+
+1. `name`: When created by COSI, the `Bucket` name is generated in this format: _"<BR.namespace*>-<BR.name*>-<uuid>"_. If an admin creates a `Bucket`, as is necessary for brownfield access, they can use any name. The uuid is unique within a cluster. `*` the first 100 characters of the namespace and the name are used.
+    cosi.io/provisioner:
+    cosi.io/bucket-prefix:
+2. `labels`: added by COSI. The "cosi.io/provisioner" key's value is the provisioner name. The "cosi.io/bucket-prefix" key's value is the `BucketRequest.bucketPrefix` value when specified. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
+3. `finalizers`: added by COSI to defer `Bucket` deletion until backend deletion ops succeed. Implementation may add one finalizer for the BR and one for the BA.
+4. `provisioner`: (optional) The provisioner field as defined in the `BucketClass`. If empty then there is no driver/sidecar, thus the admin had to create the `Bucket` and `BucketAccess` instances.
+5. `retentionPolicy`: Prescribes the outcome when the `BucketRequest` is deleted.
+   - _Retain_: (default) the `Bucket` and its data are preserved. The `Bucket` can potentially be reused.
+   - _Delete_: the bucket and its contents are destroyed.
```

Comment:
Yeah, this follows the exact same pattern as PVs.

can you explain what is OBC and what the "issue" is?

Reply:
OB/OBC are resources defined in the [object bucket library](https://github.com/kube-object-storage/lib-bucket-provisioner)

Reply:
> Yeah, this follows the exact same pattern as PVs.

So, a `retentionPolicy` = "Delete" will deleted the backend bucket [1] and, of course, the k8s resources associated with that bucket. It works this way for greenfield and brownfield.

1. it is ultimately up to each driver whether or not the backend bucket is actually deleted. For example, instead of permanently deleting the bucket, the driver could archive/rename it. etc. However, whatever the driver implements has to be idempotent and reentrant.

Reply:
> Yeah, this follows the exact same pattern as PVs.
> 
> can you explain what is OBC and what the "issue" is?

It is not an "issue" issue, I meant for brownfield cases it may have objects prior to bucket provisioning will be removed as part of `Delete` policy.